trade.lic

```diff
@@ -205,58 +662,851 @@ class Trade
     Flags.add('caravan-upgrade', /Unfortunately, this load would kill/)
     bput('give my contract to clerk', /The .* clerk accepts your contract and peruses it\./)
-    recall_caravan if Flags['caravan-upgrade']
+    if Flags['caravan-upgrade']
+      bput('rent caravan', 'You want to upgrade from a')
+      recall_caravan?
+    end
     bput("put my contract in my #{@contract_container}", 'You put')
     Flags.delete('caravan-upgrade')
   end
+
+  def deposit_coins(withdraw_amt, return_to_id)
+    return if wealth(@hub_city) < withdraw_amt
+
+    walk_to(@town_data[@hub_city]['deposit']['id'])
+    case bput('deposit all', 'you drop all your', 'You hand the clerk some coins', "You don't have any", 'There is no teller here', 'reached the maximum balance I can permit')
+    when 'There is no teller here'
+      walk_to(return_to_id)
+      return
+    end
+    minimize_coins(withdraw_amt).each { |amount| withdraw_exact_amount?(amount, @settings) }
+    walk_to(return_to_id)
+  end
+
+  def debugging?
+    UserVars.trade_debug
+  end
+
+### TRAINING SECTION ###
+
+  def attunement_routine
+    return unless @training_token == 'Attunement'
+    echo "attunement_routine" if debugging?
+    return unless @caravan_being_led || @step_toggle > 1
+    bput("perc mana", 'You reach out')
+    @step_toggle = 0
+    wipe_token
+    waitrt?
+  end
+
+  def appraisal_routine
+    return unless @training_token == 'Appraisal'
+    return unless @caravan_being_led || @step_toggle > 1
+    be_inside_caravan
+    case @appraisal_queue.first
+    when 'zills'
+      @appraisal_queue = @appraisal_queue.rotate
+      return appraisal_routine unless assess_zills?
+    when 'bundle'
+      @appraisal_queue = @appraisal_queue.rotate
+      return appraisal_routine unless appraise_bundle? 
+    when 'pouch'
+      @appraisal_queue = @appraisal_queue.rotate
+      return appraisal_routine unless appraise_worn_pouch?
+    when 'full-pouch'
+      unless appraise_held_pouch?
+        @appraisal_queue = @appraisal_queue.rotate
+        return appraisal_routine
+      end
+    end
+    @step_toggle = 0
+    wipe_token
+    @cooldown_timers['Appraisal'] = Time.now
+  end
+  
+  def appraise_worn_pouch?
+    case bput("appraise my #{@gem_pouch_adjective} pouch", 'Roundtime', 'You can.t appraise the', 'Appraise what')
+    when 'You can.t appraise the', 'Appraise what'
+      @appraisal_queue -= ['pouch']
+      false
+    end
+    true
+  end
+  
+  def appraise_held_pouch?
+    @pouch_ordinal ||= 0
+    @pouch_ordinal = 0 if @pouch_ordinal > 10
+    case bput("get #{$ORDINALS[@pouch_ordinal]} pouch from my #{@full_pouch_container}", '^You get ', '^What were you referring')
+    when /^You get /
+      bput('appraise my pouch', 'Roundtime')
+      waitrt?
+      bput("put my pouch in my #{@full_pouch_container}", 'You put')
+      @pouch_ordinal += 1
+      true
+    else
+      @appraisal_queue -= 'full-pouch' if @pouch_ordinal == 0
```

Comment:
Running into this error: 
```
 --- Lich: error: no implicit conversion of String into Array        
         trade:862:in `-'                                            
         trade:862:in `appraise_held_pouch?'                         
 --- Lich: trade has exited.                                         
```

I think this line needs to be `@appraisal_queue -= ['full-pouch'] if @pouch_ordinal == 0`

Reply:
Yep, good catch.  I'll check the others.  That's what I get for doing it in a hurry.